File: protocol/bridge/common/src/main/java/ink/reactor/protocol/bridge/common/handler/state/StatePacketHandlerStorage.java
package ink.reactor.protocol.bridge.common.handler.state;

import ink.reactor.protocol.api.connection.PlayerConnection;
import ink.reactor.protocol.api.buffer.reader.ReaderBuffer;
import ink.reactor.protocol.api.packet.PacketHandler;
import ink.reactor.protocol.api.packet.PacketHandlerStorage;
import ink.reactor.protocol.bridge.common.handler.MutablePacketHandlerStorage;
import ink.reactor.protocol.bridge.common.handler.SingleHandlerStorage;

public final class StatePacketHandlerStorage implements PacketHandlerStorage {

    private final MutablePacketHandlerStorage[] packetHandlerStorage;

    public StatePacketHandlerStorage(int size) {
        this.packetHandlerStorage = new MutablePacketHandlerStorage[size];
    }

    @Override
    public void callHandlers(final PlayerConnection connection, final ReaderBuffer buffer, final int packetID) {
        if (packetID < 0 || packetID >= packetHandlerStorage.length) {
            throw new ArrayIndexOutOfBoundsException("Packet id need be in the range 0-" + (packetHandlerStorage.length-1) + " but found " + packetID);
        }
        if (packetHandlerStorage[packetID] != null) {
            packetHandlerStorage[packetID].callHandlers(connection, buffer);
        }
    }

    @Override
    public void add(final PacketHandler handler) {
        final int packetID = handler.packetId();
        if (packetID < 0 || packetID >= packetHandlerStorage.length) {
            throw new ArrayIndexOutOfBoundsException("Packet id need be in the range 0-" + (packetHandlerStorage.length-1) + " but found " + handler.packetId());
        }

        final MutablePacketHandlerStorage mutablePacketHandlerStorages = packetHandlerStorage[packetID];
        packetHandlerStorage[packetID] = mutablePacketHandlerStorages == null
            ? new SingleHandlerStorage(handler)
            : mutablePacketHandlerStorages.add(handler);
    }

    @Override
    public void remove(final PacketHandler handler) {
        final int packetID = handler.packetId();
        if (packetID < 0 || packetID >= packetHandlerStorage.length) {
            throw new ArrayIndexOutOfBoundsException("Packet id need be in the range 0-" + (packetHandlerStorage.length-1) + " but found " + handler.packetId());
        }
        final MutablePacketHandlerStorage mutablePacketHandlerStorages = packetHandlerStorage[packetID];
        if (mutablePacketHandlerStorages != null) {
            packetHandlerStorage[packetID] = mutablePacketHandlerStorages.remove(handler);
        }
    }

    @Override
    public void add(final PacketHandler... handlers) {
        for (final PacketHandler packetHandler : handlers) {
            add(packetHandler);
        }
    }

    @Override
    public void remove(final PacketHandler... handlers) {
        for (final PacketHandler packetHandler : handlers) {
            remove(packetHandler);
        }
    }

    @Override
    public void clear(int id) {
        packetHandlerStorage[id] = null;
    }
}

File: networking/internal/src/main/kotlin/ink/reactor/network/internal/packet/storage/StatePacketHandlerStorage.kt
package ink.reactor.network.internal.packet.storage

import ink.reactor.network.api.packet.Packet
import ink.reactor.network.api.packet.handler.PacketHandler
import ink.reactor.network.api.packet.handler.PacketHandlerStorage
import ink.reactor.network.api.player.PlayerConnection

class StatePacketHandlerStorage(size: Int) : PacketHandlerStorage {

    private val packetHandlerStorage = arrayOfNulls<MutablePacketHandlerStorage>(size)

    override fun callHandlers(connection: PlayerConnection, packet: Packet) {
        val packetID = packet.id()
        requirePacketIdInRange(packetID)
        packetHandlerStorage[packetID]?.callHandlers(connection, packet)
    }

    override fun add(handler: PacketHandler) {
        val packetID = handler.packetId()
        requirePacketIdInRange(packetID)

        val currentStorage = packetHandlerStorage[packetID]
        packetHandlerStorage[packetID] = currentStorage?.add(handler) ?: SingleHandlerStorage(handler)
    }

    override fun remove(handler: PacketHandler) {
        val packetID = handler.packetId()
        requirePacketIdInRange(packetID)

        packetHandlerStorage[packetID] = packetHandlerStorage[packetID]?.remove(handler)
    }

    override fun add(vararg handlers: PacketHandler) {
        for (handler in handlers) {
            add(handler)
        }
    }

    override fun remove(vararg handlers: PacketHandler) {
        for (handler in handlers) {
            remove(handler)
        }
    }

    override fun clear(id: Int) {
        packetHandlerStorage[id] = null
    }

    private fun requirePacketIdInRange(packetID: Int) {
        if (packetID < 0 || packetID >= packetHandlerStorage.size) {
            throw ArrayIndexOutOfBoundsException(
                "Packet id needs to be in the range 0-${packetHandlerStorage.size - 1} but found $packetID"
            )
        }
    }
}
